builder/remote.go

```diff
@@ -0,0 +1,119 @@
+package builder
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"regexp"
+
+	"github.com/docker/docker/pkg/httputils"
+)
+
+// When downloading remote contexts, limit the amount (in bytes)
+// to be read from the response body in order to detect its Content-Type
+const maxPreambleLength = 100
+
+const acceptableRemoteMIME = `(?:application/(?:(?:x\-)?tar|octet\-stream|((?:x\-)?(?:gzip|bzip2?|xz)))|(?:text/plain))`
+
+var mimeRe = regexp.MustCompile(acceptableRemoteMIME)
+
+// MakeRemoteContext downloads a context from remoteURL and returns it.
+//
+// If contentTypeHandlers is non-nil, then the Content-Type header is read along with a maximum of
+// maxPreambleLength bytes from the body to help detecting the MIME type.
+// Look at acceptableRemoteMIME for more details.
+//
+// If a match is found, then the body is sent to the contentType handler and a (potentially compressed) tar stream is expected
+// to be returned. If no match is found, it is assumed the body is a tar stream (compressed or not).
+// In either case, an (assumed) tar stream is passed to MakeTarSumContext whose result is returned.
+func MakeRemoteContext(remoteURL string, contentTypeHandlers map[string]func(io.ReadCloser) (io.ReadCloser, error)) (ModifiableContext, error) {
+	f, err := httputils.Download(remoteURL)
+	if err != nil {
+		return nil, fmt.Errorf("Error downloading remote context %s: %v", remoteURL, err)
+	}
+	// not defer-closing f.Body directly, to make sure we call
+	// making sure all codepaths defer-close ReadClosers once.
+	contextReader := f.Body
+
+	if contentTypeHandlers != nil {
+		ct := f.Header.Get("Content-Type")
+		clen := f.ContentLength
+
+		defer f.Body.Close()
+		var contentType string
+		contentType, contextReader, err = inspectResponse(ct, f.Body, clen)
+		if err != nil {
+			return nil, fmt.Errorf("Error detecting content type for remote %s: %v", remoteURL, err)
+		}
+
+	handleContentType:
+		if fn, ok := contentTypeHandlers[contentType]; ok {
+			defer contextReader.Close()
+			if contextReader, err = fn(contextReader); err != nil {
+				return nil, err
+			}
+		} else if contentType != "" {
+			// use fallback handler supplied by caller (for raw tar contexts)
+			contentType = ""
+			// using goto because the defer-close in the 'if' above would make the code more complicated to read.
+			goto handleContentType
```

Comment:
I didn't understand this. Seems like you just need always close `contextReader`. I'm pretty sure that without goto will be simpler.